src/tox/config/__init__.py

```diff
@@ -176,6 +177,10 @@ def _replace_forced_dep(self, name, config):
                 return forced_dep
         return name
 
+    @staticmethod
+    def _cut_off_dep_comment(name):
+        return re.sub(r'\s+#.*', '', name).strip()
```

Comment:
probably should be a global/class level regex compile 👍 here just sub on it.

Reply:
Why? What do you mean?

Reply:
@gaborbernat is suggesting a `re.compile(...)` at the module scope so this call doesn't sometimes-compile a regex (performance)

Reply:
I followed the style of other code around  like in line 143: https://github.com/tox-dev/tox/pull/1262/files#diff-d33d19b1ff1730f9a176eed3f27fe64aR143

Reply:
@asottile I'd argue that `re` caches compiled regexes internally making such optimization useless.

Reply:
😄 we should probably fix those too I think 🤦‍♂️ but out of scope for this PR

Reply:
Anyway. This suggestion isn't exactly in the scope of this PR either. It's a further refactoring.

Reply:
https://stackoverflow.com/a/12514276 so seems it's mostly ok 👍 though implementation detail the cache

Reply:
Yep, that's what I was referring to

Reply:
it's more that this call could _sometimes_ trigger a compile side-effect slowdown, whereas the module constant will never trigger a compile side-effect

caches aren't perfect, if we know it's going to be used more than once it's better to compile and keep a strong reference ourselves and not depend on implementation detail (especially when that implementation detail is _random eviction_ in <cpython3.6!)

Reply:
Okay, it all makes sense but it's a refactoring which should be done separately

Reply:
Is it possible for environment markers to contain the comment marker? https://www.python.org/dev/peps/pep-0496/ How do we handle this case?

Reply:
@gaborbernat the comment marker is `#` and env markers use `;`. What's the problem?

Reply:
```pywin32 >=1.0 ; sys_platform == '#my-magic-platform' or sys_platform == 'your-#-platform' # so what now ``` 
https://www.python.org/dev/peps/pep-0496/#id9

Reply:
we probably want to strip to the right of the first non-quoted comment marker only

Reply:
@gaborbernat this will cut off only comments with a space in front of the hash char

Reply:
I've done it like this because URLs can also contain fragments.

Reply:
one would argue we should do it even if it does now have space, string literals inside the env marker can have space and then # too

Reply:
It's the easiest solution possible. I don't see any case where it's impossible to avoid space in front of #. That's why I'd stop here instead of making this more complicated.

Reply:
I would argue this needs to support both ``a#comment`` and ``pywin32 >=1.0 ; sys_platform == '# my-magic-platform' or sys_platform == 'your- #-platform' # so what now ``, so it needs to be more complicated.

Reply:
It'd be hard to make it work in URLs (outside quotes), then: 
git+https://github.com/manahl/pytest-plugins.git#egg=pytest-verbose-parametrize&subdirectory=pytest-verbose-parametrize.
In this case, you never know whether it's a fragment or a comment

Reply:
Sounds like we found why we never implemented this. Maybe we should reject this, @asottile?

Reply:
hmmm yeah seems like there's potential for over-zealous splitting here -- though I'm surprised that the ini parsing isn't stripping comments already :thinking: 

I think the risk of this change is pretty low though, I can't think of any _common_ dependency markers this would break

Reply:
I think it's moderately possible to implement "in quotes" check while keeping the whitespace requirement.

Oh btw I think configparser under python2 strips off those comments, but not under python 3

Reply:
tox (unfortunately) (and pytest) use an alternate config parser from `pylib` :S

Reply:
ah. so what do we do now?